File: csharp/Mmokit.Sdk.Core/InterpolationBuffer.cs
using System.Collections.Generic;

namespace Mmokit.Sdk.Core
{
    /// Optional stateful per-entity playback buffer: a Sample ring + the two
    /// orchestration footguns bundled (stale-gated push, interpolate-at-render-
    /// time). Operates ONLY on Sample — carries no game semantics. The consumer
    /// owns one buffer per entity, converts its entity -> Sample (incl. rotation
    /// rule via TryNewest), and applies the InterpolationResult to its view.
    /// Headless/bot consumers ignore this and use InterpolationCore directly.
    public sealed class InterpolationBuffer
    {
        public const int DefaultRingSize = 8;
        public const double DefaultRenderDelayMs = 100;
        public const double DefaultMaxExtrapolateMs = 50;

        readonly object _sync = new();
        readonly List<Sample> _ring = new();
        uint? _authorityEpoch;

        public int RingSize { get; }
        public double RenderDelayMs { get; }
        public double MaxExtrapolateMs { get; }
        public int Count
        {
            get
            {
                lock (_sync)
                    return _ring.Count;
            }
        }

        public uint? AuthorityEpoch
        {
            get
            {
                lock (_sync)
                    return _authorityEpoch;
            }
        }

        public InterpolationBuffer(
            int ringSize = DefaultRingSize,
            double renderDelayMs = DefaultRenderDelayMs,
            double maxExtrapolateMs = DefaultMaxExtrapolateMs)
        {
            if (ringSize <= 0)
                throw new System.ArgumentOutOfRangeException(nameof(ringSize), "Ring size must be positive.");
            if (renderDelayMs < 0)
                throw new System.ArgumentOutOfRangeException(nameof(renderDelayMs), "Render delay cannot be negative.");
            if (maxExtrapolateMs < 0)
                throw new System.ArgumentOutOfRangeException(nameof(maxExtrapolateMs), "Maximum extrapolation cannot be negative.");

            RingSize = ringSize;
            RenderDelayMs = renderDelayMs;
            MaxExtrapolateMs = maxExtrapolateMs;
        }

        /// Stale-gated append (drops frames older than the newest held).
        public SamplePushStatus Push(Sample s) => PushWithStatus(s);

        /// Epoch-aware append with an explicit outcome for callers that need
        /// to gate non-interpolated fields on the exact same decision.
        public SamplePushStatus PushWithStatus(Sample s)
        {
            lock (_sync)
                return InterpolationCore.PushSampleWithStatus(_ring, ref _authorityEpoch, s, RingSize);
        }

        /// Whether Push(s) would drop s as stale. Gate non-interpolated field
        /// snapshots (size/health/…) on this to match the position ring.
        public bool IsStale(double producedAtMs, uint? authorityEpoch = null)
        {
            lock (_sync)
                return InterpolationCore.IsStaleSample(_ring, producedAtMs, _authorityEpoch, authorityEpoch);
        }

        /// Newest sample held (for a prev-rotation fallback on stationary entities).
        public bool TryNewest(out Sample s)
        {
            lock (_sync)
            {
                if (_ring.Count > 0) { s = _ring[_ring.Count - 1]; return true; }
                s = default;
                return false;
            }
        }

        /// Interpolated render pose at the given server render time.
        public bool SampleAt(double renderTimeMs, out InterpolationResult result)
        {
            lock (_sync)
                return InterpolationCore.InterpolateRing(_ring, renderTimeMs, MaxExtrapolateMs, RenderDelayMs, out result);
        }

        /// Interpolated render pose using a dynamic delay supplied by an
        /// adaptive playback controller for this frame.
        public bool SampleAt(double renderTimeMs, double renderDelayMs, out InterpolationResult result)
        {
            if (renderDelayMs < 0)
                throw new System.ArgumentOutOfRangeException(nameof(renderDelayMs), "Render delay cannot be negative.");
            lock (_sync)
                return InterpolationCore.InterpolateRing(_ring, renderTimeMs, MaxExtrapolateMs, renderDelayMs, out result);
        }

        /// Explicitly reset playback history, for example on a fresh snapshot
        /// after reconnect. AuthorityEpoch may seed the new timeline.
        public void Reset(uint? authorityEpoch = null)
        {
            lock (_sync)
            {
                _ring.Clear();
                _authorityEpoch = authorityEpoch;
            }
        }
    }
}

File: examples/cube3d/web/sdk/_core/interpolation-buffer.ts
/**
 * Optional stateful per-entity playback buffer: a Sample ring + the two
 * orchestration footguns bundled (stale-gated push, interpolate-at-render-
 * time). Operates only on Sample — no game semantics. The consumer owns one
 * buffer per entity, converts its entity -> Sample (incl. rotation rule via
 * newest()), and applies the InterpolationResult to its view. Headless/bot
 * consumers ignore this and use interpolation-core directly.
 */
import {
  type Sample,
  type SampleRing,
  type InterpolationResult,
  type PushSampleResult,
  pushSample,
  isStaleSample,
  interpolateRing,
} from "./interpolation-core.js";

export const DEFAULT_RING_SIZE = 8;
export const DEFAULT_RENDER_DELAY_MS = 100;
export const DEFAULT_MAX_EXTRAPOLATE_MS = 50;

export interface InterpolationBufferConfig {
  ringSize?: number;
  renderDelayMs?: number;
  maxExtrapolateMs?: number;
}

export class InterpolationBuffer implements SampleRing {
  samples: Sample[] = [];
  authorityEpoch?: number;
  readonly ringSize: number;
  readonly renderDelayMs: number;
  readonly maxExtrapolateMs: number;

  constructor(cfg: InterpolationBufferConfig = {}) {
    this.ringSize = cfg.ringSize ?? DEFAULT_RING_SIZE;
    this.renderDelayMs = cfg.renderDelayMs ?? DEFAULT_RENDER_DELAY_MS;
    this.maxExtrapolateMs = cfg.maxExtrapolateMs ?? DEFAULT_MAX_EXTRAPOLATE_MS;
  }

  /** Stale-gated append (drops frames older than the newest held). */
  push(s: Sample): PushSampleResult {
    return pushSample(this, s, this.ringSize);
  }

  /** Whether push(s) would drop s as stale — gate non-position field snapshots on this. */
  isStale(producedAtMs: number, authorityEpoch?: number): boolean {
    return isStaleSample(this, producedAtMs, authorityEpoch);
  }

  /**
   * Forget samples and their authority fence. Use this only when an enclosing
   * stream generation supersedes the old stream (for example, transfer
   * rollback); an ordinary fresh snapshot should retain a compatible ring.
   */
  reset(): void {
    this.samples = [];
    this.authorityEpoch = undefined;
  }

  /** Newest sample held, or null when empty (for prev-rotation fallback). */
  newest(): Sample | null {
    return this.samples.length > 0 ? this.samples[this.samples.length - 1] : null;
  }

  /** Interpolated render pose at the given server render time, or null when empty. */
  sampleAt(
    renderTimeMs: number,
    renderDelayMs: number = this.renderDelayMs,
  ): InterpolationResult | null {
    return interpolateRing(this, renderTimeMs, this.maxExtrapolateMs, renderDelayMs);
  }
}
